### PR TITLE
Pass through query metadata from json_api_client

### PR DESCRIPTION
**Pull request by sringling**
Pass through query metadata from json_api_client even when we are returning just one result from a find

### CODE REVIEW

**Comment by bluetwin** [comment]
@sringling 
So I after thinking about this a bit more I feel like we should just use an empty response and catch the meta there instead of adding the query meta and errors onto the class. 

example of not found:

``` ruby
2.2.0 :003 >Lawyer.find(1)
  HTTP GET (11.89ms)   http://example.com:80/api/1/lawyers/1.json
  Response status   Net::HTTPNotFound (404)
  Response body   {"lawyers":[],"meta":{"errors":"RecordNotFound","status":404}}
```

example of redirect:

``` ruby
2.2.0 :003 >Lawyer.find(1).size
  HTTP GET (15.94ms)   http://example.com:80/api/1/lawyers/1.json
  Response status   Net::HTTPNotFound (404)
  Response body   {"lawyers":[],"meta":{"errors":"Redirect","status":"301", "url":"http://example.com:80/api/1/lawyers/2.json","id":"98104-oh-john-smith-2"}}
 => 0
```

thoughts?

**Comment by sringling** [comment]
I was originally hoping we could take whatever object from the response and say url_for(response_object) and end up with the redirect URL.  That didn't really work because url_for expects specific things... so I ended up writing this:

```
        when (300...399).include?(status_code)
          class_id_field_name = get_most_likely_id_for_path(query_result, request.path_parameters)
          merged_parameters = params.merge(class_id_field_name => query_result.to_param)
          redirect_target_url = url_for(merged_parameters)

          case
            when status_code == 301
              perm_redirect_to redirect_target_url
            else
              redirect_to redirect_target_url
          end
```

So your suggestion reduces our response payload and takes get_most_likely_id_for_path and turns it into response.meta.id, so that seems like a better way to go.  We can then gsub that id in our original request path and redirect to that new target?  That can be an issue for id = 1 in a nested path like website/blah/1/eehhhh/1, we would replace 1 multiple times or in the wrong place.  Maybe we can include the id param name as well?  So add id to the meta + id_param_name?  We know what it is on the server because we grab it out of the list, like params[:id] or params[:thing_id], so we can always return id_param_name = :thing_id.

Does that sound good?

File: lib/json_api_resource/queryable.rb [hunk 1]
```diff
@@ -27,6 +32,31 @@ def where(opts = {})
...
+      # we lose these things.  We want them, so we will assign them to this object.
+      def get_single_result(result_set)
+        single_result = result_set.first
+
+        single_result.methods.select{|method_name| is_query_result_setter_method(method_name)}.each do |setter_method_name|
+          getter_method_name = setter_method_name.to_s.gsub("=", "")
+          if (result_set.methods.include?(getter_method_name.to_sym))
+            single_result.send(setter_method_name, result_set.send(getter_method_name))
+          end
+        end
```
Comment (bluetwin) [inline lib/json_api_resource/queryable.rb:51]: what do you think about breaking this down a bit?

``` ruby
def get_single_result(result_set)
  query_methods.each do |setter|
    getter = setter.to_s.gsub("=", "")
    if (result_set.methods.respond_to?(getter.to_sym))
      single_result.send(setter, result_set.send(getter))
   end
  end
end
...


def query_methods(single_obj)
  methods = []
  QUERY_RESULT_METADATA_SETTERS.each do |setter|
    methods << setter if single_obj.responds_to?(setter)
  end
  methods
end


```

File: .gitignore [hunk 2]
```diff
@@ -29,3 +29,5 @@ config/secrets.yml
 /vendor/assets/bower_components
 *.bowerrc
 bower.json
+
+.idea/
```
Comment (bluetwin) [inline .gitignore:33]: silly rubymine, always trying to get in our git repos

**Pull request merged**